diff --git a/hw/display/virtio-gpu-virgl.c b/hw/display/virtio-gpu-virgl.c
index 07f6355ad6..c0f2c8eaa5 100644
--- a/hw/display/virtio-gpu-virgl.c
+++ b/hw/display/virtio-gpu-virgl.c
@@ -1437,6 +1437,10 @@ static int virtio_gpu_virgl_init(VirtIOGPU *g)
 #if VIRGL_RENDERER_CALLBACKS_VERSION >= 4
     if (qemu_egl_display) {
         virtio_gpu_3d_cbs.version = 4;
+        flags |= VIRGL_RENDERER_USE_EGL;
+        if (qemu_egl_mode == DISPLAY_GL_MODE_ES) {
+            flags |= VIRGL_RENDERER_USE_GLES;
+        }
         virtio_gpu_3d_cbs.get_egl_display = virgl_get_egl_display;
 #if VIRGL_CHECK_VERSION(1, 1, 2)
         virtio_gpu_3d_cbs.write_fence         = virgl_write_async_fence;
